GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,466 @@
+                {
+                    this.latestVersion = highestVersion;
+                }
+
+                newVersion = this.latestVersion?.Identity?.Version?.Version;
+
+                if (newVersion != null)
+                {
+                    this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - new version available: installedVersion: {this.installedVersion}, latestAvailableVersion: {highestVersion}");
+                    message = $"New version {highestVersion.Identity.Version} is available.";
+                    return true;
+                }
+                else if (highestVersion != null)
+                {
+                    this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - up-to-date");
+                    message = $"Latest available version is {highestVersion.Identity.Version}, you are up-to-date";
+                    return true;
+                }
+                else
+                {
+                    this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - no versions available from feed.");
+                    message = $"No versions available via feed.";
+                }
+            }
+            catch (Exception ex)
+            {
+                this.tracer.RelatedError($"{nameof(this.TryQueryNewestVersion)} failed with: {ex.Message}");
+                message = ex.Message;
+                newVersion = null;
+            }
+
+            return false;
+        }
+
+        public bool TryDownloadNewestVersion(out string errorMessage)
+        {
+            if (this.latestVersion == null)
+            {
+                errorMessage = "No new version to download. Query for latest version to ensure a new version is available before downloading.";
+                return false;
+            }
+
+            using (ITracer activity = this.tracer.StartActivity(nameof(this.TryDownloadNewestVersion), EventLevel.Informational))
+            {
+                try
+                {
+                    this.DownloadedPackagePath = this.nuGetFeed.DownloadPackage(this.latestVersion.Identity).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
+                {
+                    activity.RelatedError($"{nameof(this.TryDownloadNewestVersion)} - error encountered: ${ex.Message}");
+                    errorMessage = ex.Message;
+                    return false;
+                }
+            }
+
+            errorMessage = null;
+            return true;
+        }
+
+        public bool TryCleanup(out string error)
+        {
+            error = null;
+            Exception e;
+            bool success = this.fileSystem.TryDeleteDirectory(this.localUpgradeServices.TempPath, out e);
+
+            if (!success)
+            {
+                error = e?.Message;
+                this.tracer.RelatedError($"{nameof(this.TryCleanup)} - Error encountered: {error}");
```

Comment:
Trace full exception stack?

Reply:
(applies in a few spots in this class)

Reply:
> Why not just let exceptions with appropriate error messages or types flow up the call stack, and then catch them, trace them and present the human actionable message at a very high level?

We've had a few reasons for generally preferring the TryXXX pattern (@kewillford might remember some reasons):

- Control flow via TryXXX is faster than exceptions
- We've needed to know exactly when\where a failures occur to know whether it's safe or VFS4G to continue.  If multiple methods can throw the same type exception and they're called in a try { } block it's difficult to know which failed (unless each call has its own try { } block).
- Catching exceptions at a high level was making it difficult to diagnose user issues because the logs would be missing details from the lower parts of the callstack.  The TryXXX approach allows for tracing the relevant parameters/details for the specific call that failed.